### PR TITLE
Fix confusing CTC documentation

### PR DESCRIPTION
**Pull request by al-muammar**
From the docs it seems that CTC decoder have problems and it can't handle labelings like `A B B`, but if you will look at the [code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/ctc/ctc_decoder.h#L84), it is not the case, everything is OK, just typo in the docs.

### CODE REVIEW

**Comment by Jihadik** [comment]
I signed it!

**Comment by igormq** [comment]
Hi, @Jihadik.

If you try this piece of code

```python
import numpy as np
import tensorflow as tf

#A B B * B * B, where * means blank
abb_b_b = np.array([[1, 0, 0], [0, 1, 0], [0, 1, 0], [0, 0, 1], [0, 1, 0], [0, 0, 1], [0, 1, 0]]) # One hot encoding
abb_b_b = abb_b_b[None, :, :].transpose((1,0,2)) # Added batch axis and transpose (it's time major notation)

sess = tf.InteractiveSession()
merge = tf.sparse_tensor_to_dense(tf.nn.ctc_greedy_decoder(abb_b_b, seq_len)[0][0]).eval()
non_merge = tf.sparse_tensor_to_dense(tf.nn.ctc_greedy_decoder(abb_b_b, seq_len, merge_repeated=False)[0][0]).eval()

print(merge)
print(non_merge)
```

you'll get this output

```python
[[0 1 1 1]] # Which means ABBB
[[0 1 1 1 1]] # Which means ABBBB. There is a extra 1 in the docs!
```
So, the docs seems to have a problem, but not the problem that you've point to. If you look at [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/ctc/ctc_decoder.h#L88) line of code, you can see that if `merge_repeated = False`, then the greedy decoder will append to output only the maximum target if it isn't blank. In case of `merge_repeated=True`, the maximum target will only be appended to output if it isn't the blank label and if it isn't equal the previous label. 

Look at the comment [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/ctc_ops.py#L184). As you can see, there is a typo. Instead of * `A B B B B B` if `merge_repeated=False`., it should be * `A B B B B` if `merge_repeated=False`.

So, I recommend you to change your PR to fix this typo! :)

File: tensorflow/python/ops/ctc_ops.py
Comment (igormq) [inline tensorflow/python/ops/ctc_ops.py]: Could you change this to * `A B B B B` if `merge_repeated=False`?

**Review by igormq — APPROVED** [review]

**Comment by zheng-xq** [comment]
@tensorflow-jenkins test this please.

File: tensorflow/python/ops/ctc_ops.py
Comment (ebrevdo) [inline tensorflow/python/ops/ctc_ops.py]: This seems at odds with the unit test [here](https://github.com/petewarden/tensorflow_ios/blob/master/tensorflow/contrib/ctc/ctc_decoder_ops_test.py#L132), no?

Comment (ebrevdo) [inline tensorflow/python/ops/ctc_ops.py]: sorry - that was to the wrong repo, but the correct copy of the same file, [same line](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/ctc_decoder_ops_test.py#L133).

Comment (Jihadik) [inline tensorflow/python/ops/ctc_ops.py]: No, everything is fine. In those two examples, if you take `argmax` you will obtain:
1. `A * * B` which will transform into `A B`.
2. `B B * B A` which will transform into `B B A`.

Comment (Jihadik) [inline tensorflow/python/ops/ctc_ops.py]: In fact it is really confusing since from docs it seems that there is no way to obtain two similar labels in a row like `B B A` (that is one of the reasons why we need `blank`) which is not true for CTC.

**Comment by Jihadik** [comment]
@tensorflow-jenkins test this please.

**Comment by ebrevdo** [comment]
You're right, thanks!

**Comment by ebrevdo** [comment]
@tensorflow-jenkins test this please.

**Review by ebrevdo — APPROVED** [review]

**Comment by Jihadik** [comment]
Are we merging it, @ebrevdo?

**Comment by ebrevdo** [comment]
Let's pause on it.  I need to find out how many people the true fix would break.

**Review by ebrevdo — DISMISSED** [review]
Please hold on a few days.

**Comment by vrv** [comment]
Ping for @ebrevdo -- let me know approximately when you think you'll have found your answer?

**Comment by Jihadik** [comment]
Btw, what is the problem with merging comments? :)

File: tensorflow/core/ops/ops.pbtxt
Comment (vrv) [inline tensorflow/core/ops/ops.pbtxt]: This file is autogenerated, you can remove it from the PR and it'll get updated after merging sometime.

File: tensorflow/g3doc/api_docs/python/functions_and_classes/shard6/tf.nn.ctc_greedy_decoder.md
Comment (vrv) [inline tensorflow/g3doc/api_docs/python/functions_and_classes/shard6/tf.nn.ctc_greedy_decoder.md]: same with this one and nn.md

**Comment by Jihadik** [comment]
As I thought :), thanks!

**Comment by Jihadik** [comment]
@vrv, I fixed it!

**Comment by vrv** [comment]
(Thanks, now will wait for @ebrevdo to review)

**Comment by Jihadik** [comment]
@ebrevdo, ping!

**Comment by ebrevdo** [comment]
It's not the documentation that's wrong - it's the code.  Would you like to fix the code instead?

**Comment by Jihadik** [comment]
No problems, I will look at the code a little bit later.

**Comment by lingz** [comment]
Can confirm that I ran into this the other day. I was doing some training with labels that had repeated symbols, but it would output the warning "No Valid Path Found", and would return a cost of infinity.

**Comment by ebrevdo** [comment]
Ling, I'm not sure that is related.

**Comment by lingz** [comment]
Actually yeah, I investigated further and it was just that I did not make the seq_len long enough to allow for the blanks to be between the labels, so thus "No Valid Path Found"

Though I'm confused about your earlier comment @ebrevdo where you said it's not the documentation but the code that is wrong. The unit test linked and @igormq seems to show merge_repeated demonstrating the correct behavior?

**Comment by ebrevdo** [comment]
The unit test is either wrong, or doesn't actually test repeats correctly.

**Comment by martinwicke** [comment]
Ping! What's the next step for this?

**Comment by Jihadik** [comment]
Soon I will try to fix it. I was really busy last few weeks.

**Comment by saseptim** [comment]
I have run into several issues in the CTC decoding.

1) The greedy decoder seems to work right but has a problem with the documentation (still). Documentation online still shows 
A B B * B * B (where '*' is the blank label) becomes
    A B if merge_repeated=True.
    A B B B B B if merge_repeated=False.

Both results are wrong.
This should read:
    A B B B if merge_repeated=True.
    A B B B B  if merge_repeated=False.

2) I had a problem with the beamsearch decoder. Running the code of **igormq** above, but switching in the beamsearch decoder in place of the greedy decoder via:
`merge = tf.sparse_tensor_to_dense(tf.nn.ctc_beam_search_decoder(abb_b_b, seq_len, merge_repeated=False)[0][0]).eval()

non_merge = tf.sparse_tensor_to_dense(tf.nn.ctc_beam_search_decoder(abb_b_b, seq_len, merge_repeated=False)[0][0]).eval()

print(merge)
print(non_merge)
`

the results for both are:
[[0 1 0 1]] or A B A B

which makes no sense.

Additionally, I have also had the problem where the beamsearch decoder did not properly deal with repeated characters and had to switch merge_repeated to False. 

In summary, documentation problems in greedy decoder,  code problems in beamsearch decoder.

**Comment by ebrevdo** [comment]
You have merge_repeated=false in both cases for your beam search code. For
greedy decoder, the documentation isn't wrong, the code is. I'll look into
the beam search decoder in early January.

**Comment by saseptim** [comment]
Regarding the beamsearch, you are right, I copied the code wrong (now it is fixed) but the results are the same - [[0 1 0 1]] - still incorrect.

As for the greedy decoder, I do not understand why you think the code is wrong. The code acts for me as expected. The comments, on the other hand, are certainly wrong, just the fact that the result has 5 B's and the input has 4.

**Comment by ebrevdo** [comment]
For greedy beam search, the code should act like the documentation... Not
the other way around. It's a bug.

On Dec 13, 2016 11:12 PM, "saseptim" <notifications@github.com> wrote:

Regarding the beamsearch, you are right, I copied the code wrong (now it is
fixed) but the results are the same - [[0 1 0 1]] - still incorrect.

As for the greedy decoder, I do not understand why you think the code is
wrong. The code acts for me as expected. The comments, on the other hand,
are certainly wrong, just the fact that the result has 5 B's and the input
has 4.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/tensorflow/tensorflow/pull/5194#issuecomment-266981939>,
or mute the thread

**Comment by Jihadik** [comment]
@ebrevdo, I really disagree with you. If `merge_repeated=True` then it is exactly [Graves' CTC](ftp://ftp.idsia.ch/pub/juergen/icml2006.pdf) and in case `A B B * B * B` firstly it merges repeated labels and make `A B * B * B` then it removes all blanks and our sequence become `A B B B`, not `A B` as in docs now. If it firstly remove all blanks and the merge labels then we will never have two consecutive labels in a row which is not good and was not inteded by Graves.

If I'm wrong, please prove it.

**Comment by saseptim** [comment]
I agree, there is a problem with the documentation of the greedy decoder and the code of the beamsearch decoder.

**Comment by drpngx** [comment]
@Jihadik I really appreciate your willingness to contribute. Basically, the way it works is that google staff is maintaining the code and needs to make the final decision on changes we're comfortable with. We can't really merge @ebrevdo has a different opinion.

**Comment by ebrevdo** [comment]
I'll revisit tomorrow.

**Comment by ebrevdo** [comment]
@Jihadik, you are right.  lgtm.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Review by rmlarsen — APPROVED** [review]

**Pull request merged**
